### DISCUSSION TITLE
Using ransackable_attributes with Active Admin errors  when using Active Storage

### DISCUSSION THREAD

**Discussion by vic778** [activeadmin/activeadmin#8077] [Help] [answered]
I'm using active admin gem for the admin dashboard. I have never faced same the ransackable error since I've used it, I have the user model which has the active storage so I got this error

```
Define a `ransackable_attributes` class method in your `ActiveStorage::Attachment` model
```

, yet rails don't create the active storage class in the model folder. I've done research about how to fix that but I didn't find a proper way.  
So as I said here is what the gem error shows like a solution 

```
class ActiveStorage::Attachment < ApplicationRecord
  def self.ransackable_attributes(auth_object = nil)
    ["blob_id", "created_at", "id", "name", "record_id", "record_type"]
  end
end
```

but rails don't create the active storage class.

**Comment by Lordnibbler** [comment]
Similar error here.  I have a model, simplified for brevity

```ruby
class Blog::Post < ApplicationRecord
  has_one_attached :featured_image
  has_many_attached :images
end
```

and when loading the ActiveAdmin views for this model I receive this error:

```
RuntimeError at /admin/blog_posts

Ransack needs Blog::Post attributes explicitly allowlisted as
searchable. Define a `ransackable_attributes` class method in your `Blog::Post`
model, watching out for items you DON'T want searchable (for
example, `encrypted_password`, `password_reset_token`, `owner` or
other sensitive information). You can use the following as a base:
```

```ruby
class Blog::Post < ApplicationRecord
  # ...
  def self.ransackable_attributes(auth_object = nil)
    ["body", "created_at", "description", "id", "id_value", "published", "slug", "title", "updated_at", "user_id"]
  end
  # ...
end
```

When updating the model accordingly, I see a new error:

```
Ransack needs Blog::Post associations explicitly allowlisted as
searchable. Define a `ransackable_associations` class method in your `Blog::Post`
model, watching out for items you DON'T want searchable (for
example, `encrypted_password`, `password_reset_token`, `owner` or
other sensitive information). You can use the following as a base:
```

```ruby
class Blog::Post < ApplicationRecord
  # ...
  def self.ransackable_associations(auth_object = nil)
    ["featured_image_attachment", "featured_image_blob", "images_attachments", "images_blobs", "user"]
  end
  # ...
end
```


And finally, when updating the model accordingly again, I'm back to (a variation) on the first error:

```
Ransack needs ActiveStorage::Attachment attributes explicitly allowlisted as
searchable. Define a `ransackable_attributes` class method in your `ActiveStorage::Attachment`
model, watching out for items you DON'T want searchable (for
example, `encrypted_password`, `password_reset_token`, `owner` or
other sensitive information). You can use the following as a base:
```

```ruby
class ActiveStorage::Attachment < ApplicationRecord
  # ...
  def self.ransackable_attributes(auth_object = nil)
    ["blob_id", "created_at", "id", "id_value", "name", "record_id", "record_type"]
  end
  # ...
end
```

**Comment by Eddayy** [comment, marked as answer]
I was also having the same issue, i found this works for me, you'll have to remove the activestorage tables from being searchable

```
ActiveAdmin.register Post do
  remove_filter :featured_image_attachment, :featured_image_blob
end
 ```

**Reply by javierjulio** [reply]
Thanks! I ran into this locally when updating the ActiveAdmin demo repository. This is a simple workaround that others should use.

**Reply by mabeller** [reply]
And for reference, if you have a has_many_attached relationship it'll need to be `:featured_images_attachments, :featured_images_blobs`

**Reply by LRosdobutko** [reply]
Do you mind if I ask what file you are supposed to insert that code snippet into? I'm fairly new to rails and basically teaching myself when I can, where I can, and some hand holding is greatly appreciated. lol

**Comment by TeamRainless** [comment]
This should be on the main page somewhere if it isn't already... I would imagine every single ActiveAdmin user (or most of them) must run into this problem eventually. Took me all day to find this discussion.

**Comment by abhinaykumar** [comment]
Took me couple of hours to realise that [this answer](https://github.com/activeadmin/activeadmin/discussions/8077#discussioncomment-7585251) points to `*_attachment` and `*_blob` version of the association on the model.

**Comment by TeamRainless** [comment]
That means you're just as smart as everyone else here... only slower! :)

**Comment by TeamRainless** [comment]
Which code snippet are you referring to?

If you mean the "has_many" relationships... They go into your model.

**Comment by innocentamadi** [comment]
Here's what worked for me: Create a module in an initializer that extends `ActiveSupport::Concern` and include the ransackable_attributes mapping.

```
# config/initializers/attachment_extensions.rb

module AttachmentExtensions
  extend ActiveSupport::Concern

  included do
    def self.ransackable_associations(*, **) = reflections.keys
    def self.ransackable_attributes(*, **) = attribute_names
  end
end

Rails.application.config.to_prepare do
  ActiveStorage::Attachment.include AttachmentExtensions
end
```

**Reply by changmason** [reply]
Confirmed this solution works, thank you so much!!

**Comment by TeamRainless** [comment]
It's not REALLY a "solution"... It's more of a workaround. In my opinion
this issue shouldn't have been closed until the issue was actually fixed. I
mean it's been YEARS now...
